parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoSchemaConverter.java

```diff
@@ -99,9 +139,9 @@ private Type.Repetition getRepetition(FieldDescriptor descriptor) {
     }
   }
 
-  private <T> Builder<? extends Builder<?, GroupBuilder<T>>, GroupBuilder<T>> addField(FieldDescriptor descriptor, final GroupBuilder<T> builder) {
+  private <T> Builder<? extends Builder<?, GroupBuilder<T>>, GroupBuilder<T>> addField(FieldDescriptor descriptor, final GroupBuilder<T> builder, ImmutableSetMultimap<String, Integer> seen, int depth) {
```

Comment:
would it make sense to consolidate seen and depth into a single data-structure that can be passed through and abstract some of the direct access to the multimap?

Reply:
?

the seen map does encode the depth as a single datastructure.  the depth being a separate arg is important b/c it's the current depth in the traversal, and is used to update the seen data structure.

Reply:
right, I was thinking of encapsulating this logic into its own class, so they can be recorded and updated together, to  1.  Reduce additional parameters that have to be passed through.
2.  Encapsulate the logic behind more mnemonic method names (e.g. AddRecursiveStep())

Reply:
i'm not sure encapsulation helps with readability or protection in this case.  they are really tracking different things, and should be understood by readers of the traversal code to know how each piece of state is used.